### DISCUSSION TITLE
关于协议中WebSocket连接方式中自定义请求头存在的问题

### DISCUSSION THREAD

**Discussion by A-kirami** [howmanybots/onebot#159] [想法 / 建议]
我目前尝试在浏览器中实现 onebot 协议（网页交互），但是在与 onebot 应用端（nonebot2）连接时出现了问题。
自定义的请求头无法传递给应用端，从而导致了出现 `Missing X-Self-ID Header` 错误。
我在检索了相关问题之后，确认到可能是浏览器中的 ws 连接不支持自定义请求头，因此我想问一下是否有可能对 onebot12 协议的 ws 连接部分进行修改以支持浏览器实现？

部分参考链接：
[Why can't websockets support custom headers? - StackExchange](https://security.stackexchange.com/questions/174349/why-cant-websockets-support-custom-headers)
[Custom headers for WebSocket JS - StackOverflow](https://stackoverflow.com/questions/4361173/http-headers-in-websockets-client-api)
[Websocket connection with authorization header - StackOverflow](https://stackoverflow.com/questions/64934740/websocket-connection-with-authorization-header)
[How else can I set custom headers? #467 - github.com/websockets/ws](https://github.com/websockets/ws/issues/467)

**Comment by ExerciseBook** [comment]
好像刚刚在群里看到了这个问题（？）

我的疑问是，你的网页和你的服务之间有没有考虑过 cookie 或者是其他的更加 web 一点的方案？

**Reply by A-kirami** [reply]
我是按照onebot协议来实现的，对于ws连接，文档中并没有给出其他的建立连接方式（可能看漏？）

**Comment by ExerciseBook** [comment]
确认需求：

1. 使用浏览器 js 制作一个 OneBot Server 实现
2. 将浏览器 js OneBot Server 使用反向 ws 功能连接到 NoneBot （或者是其他的反向代理 OneBot Client 实现）
3. 用浏览器里的一个虚拟的聊天窗口界面来和自己的机器人进行一个测试

**Reply by zhzhongshi** [reply]
这是用浏览器实现个OneBot？

**Reply by ExerciseBook** [reply]
> 

我在群里问了一下楼主，他说是的

**Comment by ayakasuki** [comment]
关于这个我只写了基于obv11部分实现在js文件里支持反向连接机器人，其中支持了一些嵌套信息，没有语音支持，有渲染图片支持，无视频支持，无md渲染支持 其中返回的是模拟的QQ号码和虚拟的无人群聊号码返回去，老实说没碰到请求头相关的问题，我不知道有没有帮助，如果有帮助的话 先放到了这里https://github.com/ayakasuki/web-onebot-link 我这个是用来给hexo连接机器人的一些chatgpt等AI智能体聊天用的。

<img width="1184" height="1149" alt="Image" src="https://github.com/user-attachments/assets/780f185d-7379-4fba-b44c-d831b7634aed" />

<img width="1123" height="1180" alt="Image" src="https://github.com/user-attachments/assets/0d6fe868-fc28-477b-8598-ccf9f1547eaa" />
